### ISSUE TITLE
WebAuthn don't work on alternate domains

### ISSUE THREAD

**Issue by yingziwu** [tootsuite/mastodon#15600]
### Expected behaviour

WebAuthn work well on both local domain and alternate domains.

### Actual behaviour

WebAuthn only work on local domain (bgme.me), don't work on alternate domains (bgme.bid, c.bgme.bid).

![0](https://user-images.githubusercontent.com/22741471/105119897-2d23fb00-5b0c-11eb-8b83-ef061316797c.png)
![1](https://user-images.githubusercontent.com/22741471/105119907-301eeb80-5b0c-11eb-8255-039c06bc6dab.png)

### Steps to reproduce the problem

Register the secureity key and log in on alternate domains.

### Specifications

Mastodon version: v3.3.0

**Comment by ThibG** [comment]
What do you mean by “alternate domain”? As far as I know, the `ALTERNATE_DOMAINS` env var is used so that people can find you by querying multiple domains, but you still have one “canonical” address and domain.

**Comment by yingziwu** [comment]
> What do you mean by “alternate domain”? As far as I know, the `ALTERNATE_DOMAINS` env var is used so that people can find you by querying multiple domains, but you still have one “canonical” address and domain.

In my case, I can log in via WebAuthn on bgme.me, but can't log in on bgme.bid.

**Comment by ThibG** [comment]
ok, this configuration isn't officially supported. I wonder what we could change to make it work though

**Comment by nightpool** [comment]
my guess is that there's no way to support this in the WebAuthn spec, since
ideas like "trusted groups of FQDNs" are not very common in web specs

**Comment by jdohm** [comment]
I don't think the expected behavior of @yingziwu is the expected behavior for WebAuthn.
If this would be even possible this would mean that WebAuthn wasn't save. Since it tries to secure the user from fishing attacks by only ever giving the "password" if the address requesting it is right. I'm **not** an expert in this field, so if there is one around, please correct me if I'm wrong.

Maybe this could be achieved by always sending the request from one address, even if the user is surfing on a different address.

**Comment by KNTRO** [comment]
Hello,

I own two U2F/WebAuthn FIDO2 security keys and none of them work with mastodon.social nor mas.to. After touching the button/ring in the key, nothing else happens.

Of, course, they **do** work in a bunch of other websites. 😉

![Security_key](https://user-images.githubusercontent.com/2568821/106440343-83bbfe00-6457-11eb-92e4-333639f3cbbd.gif)

**Comment by truongnmt** [comment]
You may have a look at this, might suit your need

https://github.com/cedarcode/webauthn-ruby/blob/master/docs/advanced_configuration.md
> If you have a multi-tenant application, or any application segmenation, where your users register and authenticate to each of these tenants or segments individuallly using different hostnames, or with different security needs, you need to go through Instance Based Configuration.
>
> However, if your application is served for just one hostname, or else if your users authenticate to only one subdmain (e.g. your application serves www.example.com and admin.example.com but all you users authenticate through auth.example.com) you can still rely on one Global Configuration.

**Comment by nightpool** [comment]
@truongnmt I don't think that will work—the page you linked is about having multiple *independent* credentials for different domains, not for using the same credential setup across my multiple domains. If we went down that route, we'd have to have the user set up the credential N times for each of N alternate domains, which is a bad idea as, again, mastodon does not expect you to be able to actually access the web interface on ALTERNATE_DOMAINS, it's only designed for webfinger

**Comment by My1** [comment]
working as expected I would say.
